README.md
TankWar
=======

Introduction:
-------------

This is a simple TankWar game written with Java. This project is basicly
written with toy codes and mainly for Java practice.


CHANGES
CHANGES:
========

Versions:
---------

Version 1.9 (2016-02-28)
------------------------

- Feature: Add multiple enemy tanks
- Feature: My tank can be hit by enemy tanks

Version 1.1 (2016-02-28)
------------------------

- Feature: add missile for tank
- Feature: add barrel for tank
- Feature: add count for missiles
- Bugfix:  remove missiles that will never be used again
- Bugfix:  restrict tank in visual area


Version 1.0 (2016-02-28)
------------------------

- Feature: we can control the direction of the tank.
- Feature: new tank!


src/com/zxjsdp/www/Missile.java
package com.zxjsdp.www;

import java.awt.*;
import java.util.List;

/**
 * Created by Jin on 2016/2/28.
 */
public class Missile {
    public static final int MISSILE_WIDTHS = 10;
    public static final int MISSILE_HEIGHT = 10;
    public static final int X_SPEED = 12;
    public static final int Y_SPEED = 12;
    int x, y;
    Tank.Direction dir;
    TankClient tc = null;
    private boolean live = true;

    public boolean isGood() {
        return good;
    }

    private boolean good;

    public Missile(int x, int y, Tank.Direction dir) {
        this.x = x;
        this.y = y;
        this.dir = dir;
    }

    public Missile(int x, int y, boolean good, Tank.Direction dir, TankClient tc) {
        this(x, y, dir);
        this.good = good;
        this.tc = tc;
    }

    public void draw(Graphics g) {
        if (!this.live) {
            tc.missiles.remove(this);
        }
        Color c = g.getColor();
        g.setColor(Color.black);
        g.fillOval(x, y, MISSILE_WIDTHS, MISSILE_HEIGHT);
        g.setColor(c);

        move();
    }

    private void move() {
        switch (dir) {
            case L:
                x -= X_SPEED;
                break;
            case LU:
                x -= X_SPEED;
                y -= Y_SPEED;
                break;
            case U:
                y -= Y_SPEED;
                break;
            case RU:
                x += X_SPEED;
                y -= Y_SPEED;
                break;
            case R:
                x += X_SPEED;
                break;
            case RD:
                x += X_SPEED;
                y += Y_SPEED;
                break;
            case D:
                y += Y_SPEED;
                break;
            case LD:
                x -= X_SPEED;
                y += Y_SPEED;
                break;
        }

        if (x < 0 || y < 0 || x > TankClient.GAME_WIDTH || y > TankClient.GAME_HEIGHT) {
            live = false;
        }
    }

    public boolean isLive() {
        return live;
    }

    public Rectangle getRect() {
        return new Rectangle(x, y, MISSILE_WIDTHS, MISSILE_HEIGHT);
    }

    public boolean hitTank(Tank t) {
        if (this.isLive() && this.getRect().intersects(t.getRect()) && t.isLive() && this.good != t.isGood()) {
            t.setLive(false);
            this.live = false;
            Explode e = new Explode(x, y, tc);
            tc.explodes.add(e);
            return true;
        }
        return false;
    }

    public boolean hitTanks(List<Tank> tanks) {
        for (int i=0; i<tanks.size(); i++) {
            if (this.hitTank(tc.tanks.get(i)))
                return true;
        }
        return false;
    }
}


src/com/zxjsdp/www/TankClient.java
package com.zxjsdp.www;

import java.awt.*;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.util.*;
import java.util.List;


/**
 * Created by Jin on 2016/2/27.
 */

public class TankClient extends Frame {
    public static final int WINDOW_INIT_POSITION = 100;
    public static final int GAME_WIDTH = 1400;
    public static final int GAME_HEIGHT = 900;
    public static final int REFRESH_TIME = 50;

    Tank myTank = new Tank(80, 80, true, Tank.Direction.STOP, this);

    List<Tank> tanks = new ArrayList<>();
    List<Explode> explodes = new ArrayList<>();
    List<Missile> missiles = new ArrayList<>();

    Image offScreenImage = null;

    public void launchFrame() {
        for (int i=0; i<10; i++) {
            tanks.add(new Tank(50 + 40 * (i+1), 100, false, Tank.Direction.D, this));
        }
        this.setLocation(WINDOW_INIT_POSITION, WINDOW_INIT_POSITION);
        this.setSize(GAME_WIDTH, GAME_HEIGHT);
        this.setTitle("TankWar");
        this.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                System.exit(0);
            }
        });
        this.setResizable(false);
        this.setBackground(Color.green);

        this.addKeyListener(new KeyMonitor());
        setVisible(true);

        new Thread(new PaintThread()).start();
    }

    @Override
    public void paint(Graphics g) {
        g.drawString("Missiles count: " + missiles.size(), 30, 60);
        g.drawString("Explodes count: " + explodes.size(), 30, 80);
        g.drawString("Enemy Tankl count: " + tanks.size(), 30, 100);

        for (int i=0; i<missiles.size(); i++) {
            Missile m = null;
            if (missiles.get(i) != null) {
                m = missiles.get(i);
            }
            m.hitTanks(tanks);
            m.hitTank(myTank);
            m.draw(g);
        }

        for (int i=0; i<explodes.size(); i++) {
            Explode e = explodes.get(i);
            e.draw(g);
        }

        for (int i=0; i<tanks.size(); i++) {
            Tank t = tanks.get(i);
            t.draw(g);
        }
        myTank.draw(g);
    }

    @Override
    public void update(Graphics g) {
        if (offScreenImage == null) {
            offScreenImage = this.createImage(GAME_WIDTH, GAME_HEIGHT);
        }
        Graphics gOffScreen = offScreenImage.getGraphics();
        Color c = gOffScreen.getColor();
        gOffScreen.setColor(Color.GREEN);
        gOffScreen.fillRect(0, 0, GAME_WIDTH, GAME_HEIGHT);
        gOffScreen.setColor(c);
        paint(gOffScreen);
        g.drawImage(offScreenImage, 0, 0, null);
    }

    private class PaintThread implements Runnable {
        public void run() {
            while (true) {
                repaint();
                try {
                    Thread.sleep(REFRESH_TIME);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private class KeyMonitor extends KeyAdapter {
        @Override
        public void keyPressed(KeyEvent e) {
            myTank.keyPressed(e);
        }

        @Override
        public void keyReleased(KeyEvent e) {
            myTank.KeyReleased(e);
        }
    }
}


src/com/zxjsdp/www/Tank.java
package com.zxjsdp.www;

import javax.xml.crypto.dsig.spec.XSLTTransformParameterSpec;
import java.awt.*;
import java.awt.event.KeyEvent;
import java.util.Random;

/**
 * Created by Jin on 2016/2/28.
 */
public class Tank {
    public static final int TANK_WIDTHS = 30;
    public static final int TANK_HEIGHT = 30;
    private static final int X_SPEED = 7;
    private static final int Y_SPEED = 7;
    private static final int MAXIMUM_STEP = 15;
    private static final int FIRE_FREQUENCY = 100;
    TankClient tc = null;
    int x, y;

    private boolean bL = false, bU = false, bR = false, bD = false;

    public boolean isGood() {
        return good;
    }

    private boolean good;

    private static Random r = new Random();

    public void setLive(boolean live) {
        this.live = live;
    }

    public boolean isLive() {
        return live;
    }

    private boolean live = true;

    enum Direction {L, LU, U, RU, R, RD, D, LD, STOP}

    private Direction dir = Direction.STOP;
    private Direction ptDir = Direction.D;

    private int step = r.nextInt(MAXIMUM_STEP) + 1;

    public Tank(int x, int y, boolean good) {
        this.x = x;
        this.y = y;
        this.good = good;
    }

    public Tank(int x, int y, boolean good, Direction dir, TankClient tc) {
        this(x, y, good);
        this.dir = dir;
        this.tc = tc;
    }

    public void draw(Graphics g) {
        if (!live) {
            if (!this.isGood()) {
                tc.tanks.remove(this);
            } else if(this.isGood()) {
                this.live = false;
            }
            return;
        }
        Color c = g.getColor();
        if (good) g.setColor(Color.red);
        else g.setColor(Color.blue);
        g.fillOval(x, y, TANK_WIDTHS, TANK_HEIGHT);
        g.setColor(c);

        switch (ptDir) {
            case L:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x, y + Tank.TANK_HEIGHT / 2);
                break;
            case LU:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x, y);
                break;
            case U:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x + Tank.TANK_WIDTHS/2, y);
                break;
            case RU:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x + Tank.TANK_WIDTHS, y);
                break;
            case R:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x + Tank.TANK_WIDTHS, y + Tank.TANK_HEIGHT/2);
                break;
            case RD:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x + Tank.TANK_WIDTHS, y + Tank.TANK_HEIGHT);
                break;
            case D:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x + Tank.TANK_WIDTHS/2, y + Tank.TANK_HEIGHT);
                break;
            case LD:
                g.drawLine(x + Tank.TANK_WIDTHS / 2, y + Tank.TANK_HEIGHT / 2, x, y + Tank.TANK_HEIGHT);
                break;
        }

        move();
    }

    void move() {
        switch (dir) {
            case L:
                x -= X_SPEED;
                break;
            case LU:
                x -= X_SPEED;
                y -= Y_SPEED;
                break;
            case U:
                y -= Y_SPEED;
                break;
            case RU:
                x += X_SPEED;
                y -= Y_SPEED;
                break;
            case R:
                x += X_SPEED;
                break;
            case RD:
                x += X_SPEED;
                y += Y_SPEED;
                break;
            case D:
                y += Y_SPEED;
                break;
            case LD:
                x -= X_SPEED;
                y += Y_SPEED;
                break;
            case STOP:
                break;
        }

        if (this.dir != Direction.STOP) {
            this.ptDir = dir;
        }

        if (x < 0) x = 0;
        if (y < 30) y = 30;
        if (x + Tank.TANK_WIDTHS > TankClient.GAME_WIDTH) x = TankClient.GAME_WIDTH - Tank.TANK_WIDTHS;
        if (y + Tank.TANK_HEIGHT > TankClient.GAME_HEIGHT) y = TankClient.GAME_HEIGHT - Tank.TANK_HEIGHT;

        if (!good) {
            if (step == 0) {
                step = r.nextInt(MAXIMUM_STEP) + 1;
                Direction[] dirs = Direction.values();
                int rn = r.nextInt(dirs.length);
                this.dir = dirs[rn];
            }
            step--;
            if (r.nextInt(FIRE_FREQUENCY) > 96)
                this.fire();
        }
    }

    public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        switch (key) {
            case KeyEvent.VK_CONTROL:
                fire();
                break;
            case KeyEvent.VK_LEFT:
                bL = true;
                break;
            case KeyEvent.VK_UP:
                bU = true;
                break;
            case KeyEvent.VK_RIGHT:
                bR = true;
                break;
            case KeyEvent.VK_DOWN:
                bD = true;
                break;
        }
        locateDirection();
    }

    public void KeyReleased(KeyEvent e) {
        int key = e.getKeyCode();
        switch (key) {
            case KeyEvent.VK_LEFT:
                bL = false;
                break;
            case KeyEvent.VK_UP:
                bU = false;
                break;
            case KeyEvent.VK_RIGHT:
                bR = false;
                break;
            case KeyEvent.VK_DOWN:
                bD = false;
                break;
        }
        locateDirection();
    }

    void locateDirection() {
        if (bL && !bU && !bR && !bD) dir = Direction.L;
        else if (bL && bU && !bR && !bD) dir = Direction.LU;
        else if (!bL && bU && !bR && !bD) dir = Direction.U;
        else if (!bL && bU && bR && !bD) dir = Direction.RU;
        else if (!bL && !bU && bR && !bD) dir = Direction.R;
        else if (!bL && !bU && bR && bD) dir = Direction.RD;
        else if (!bL && !bU && !bR && bD) dir = Direction.D;
        else if (bL && !bU && !bR && bD) dir = Direction.LD;
    }

    public Missile fire() {
        if (!this.isLive()) return null;
        int x = this.x + Tank.TANK_WIDTHS / 2 - Missile.MISSILE_WIDTHS / 2;
        int y = this.y + Tank.TANK_HEIGHT / 2 - Missile.MISSILE_HEIGHT / 2;
        Missile m = new Missile(x, y, this.good, ptDir, tc);
        tc.missiles.add(m);
        return m;
    }

    public Rectangle getRect() {
        return new Rectangle(x, y, TANK_WIDTHS, TANK_HEIGHT);
    }

}
